### PR TITLE
Replaced deprecated pkg_resources.packaging with packaging module

### PR DESCRIPTION
**Pull request by vfdev-5**
Usage of `from pkg_resources import packaging` leads to a deprecation warning:
```
DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
```
and in strict tests where warnings are errors, this leads to CI breaks, e.g.: https://github.com/pytorch/vision/pull/8092

Replacing `pkg_resources.package` with `package` as it is now a pytorch dependency: 
https://github.com/pytorch/pytorch/blob/fa9045a8725214c05ae4dcec5a855820b861155e/requirements.txt#L19

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

**Comment by vfdev-5** [comment]
This PR leads to a break as installing nightly pytorch does not install somehow `packaging`:
https://github.com/pytorch/vision/actions/runs/6783326793/job/18437544348?pr=8092
```
 Installing collected packages: mpmath, typing-extensions, sympy, networkx, MarkupSafe, fsspec, filelock, jinja2, torch
  Successfully installed MarkupSafe-2.1.2 filelock-3.9.0 fsspec-2023.4.0 jinja2-3.1.2 mpmath-1.2.1 networkx-3.0rc1 sympy-1.11.1 torch-2.2.0.dev20231107+cpu typing-extensions-4.4.0

...
Install third party dependencies prior to TorchVision install
  + python setup.py egg_info
  /opt/conda/envs/ci/lib/python3.8/site-packages/torch/nn/modules/transformer.py:20: UserWarning: Failed to initialize NumPy: numpy.core.multiarray failed to import (Triggered internally at ../torch/csrc/utils/tensor_numpy.cpp:84.)
    device: torch.device = torch.device(torch._C._get_default_device()),  # torch.device('cpu'),
  Traceback (most recent call last):
  setup.py:10: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
    from pkg_resources import DistributionNotFound, get_distribution, parse_version
  Traceback (most recent call last):
    File "setup.py", line 12, in <module>
      from torch.utils.cpp_extension import BuildExtension, CppExtension, CUDA_HOME, CUDAExtension
    File "/opt/conda/envs/ci/lib/python3.8/site-packages/torch/utils/cpp_extension.py", line 28, in <module>
      import packaging.version
  ModuleNotFoundError: No module named 'packaging.version'
```

**Comment by pmeier** [comment]
`packaging` is not installed, since it is not a dependency of `torch`.

**Comment by pmeier** [comment]
Forward fix in #113154.

**Comment by mthrok** [comment]
@vfdev-5 Can we fallback to the old import with `try expcet` ?

**Comment by vfdev-5** [comment]
OK, let's try that as another workaround

**Comment by atalman** [comment]
@vfdev-5 should we revert the PR and land it with all forward fixes when there is a workaround ?

**Comment by vfdev-5** [comment]
@atalman how about https://github.com/pytorch/pytorch/pull/113154 ? If this is a way to go, then yes, let's revert this PR and reland with suggested workarounds

**Comment by atalman** [comment]
@vfdev-5 Nightlies have been broken already for 2 days. The forward fix is  not landed yet, it may or may not fix the nightly issue. I suggest we revert this PR and then land it together with forward fixes and some domain build log demonstrating that it does not breaks the domains.

**Comment by vfdev-5** [comment]
OK, sounds good

File: torch/utils/cpp_extension.py [hunk 1]
```diff
@@ -1448,8 +1449,10 @@ def build_precompile_header(pch_cmd):
         except subprocess.CalledProcessError as e:
             raise RuntimeError(f"Compile PreCompile Header fail, command: {pch_cmd}") from e
 
-    extra_cflags_str = listToString(extra_cflags)
-    extra_include_paths_str = " ".join([f'-I{include}' for include in extra_include_paths])
+    extra_cflags_str = listToString(extra_cflags)    
+    extra_include_paths_str = (
```
Comment (vfdev-5) [inline torch/utils/cpp_extension.py:1453]: This is a drive-by fix of the failing test:
```
test/test_cpp_extensions_jit.py::TestCppExtensionJIT::test_gen_extension_h_pch
```
when `extra_include_paths = None`. Error message:
```
  File "/pytorch/torch/utils/cpp_extension.py", line 1595, in load_inline
    _check_and_build_extension_h_precompiler_headers(extra_cflags, extra_include_paths)
  File "/pytorch/torch/utils/cpp_extension.py", line 1452, in _check_and_build_extension_h_precompiler_headers
    extra_include_paths_str = " ".join([f'-I{include}' for include in extra_include_paths])
TypeError: 'NoneType' object is not iterable
```

File: torch/utils/cpp_extension.py [hunk 2]
```diff
@@ -1449,7 +1450,9 @@ def build_precompile_header(pch_cmd):
             raise RuntimeError(f"Compile PreCompile Header fail, command: {pch_cmd}") from e
 
     extra_cflags_str = listToString(extra_cflags)
-    extra_include_paths_str = " ".join([f'-I{include}' for include in extra_include_paths])
+    extra_include_paths_str = (
+        "" if extra_include_paths is None else " ".join([f"-I{include}" for include in extra_include_paths])
+    )
```
Comment (malfet) [inline torch/utils/cpp_extension.py:1455]: Nit (this looks more compact, as `" ".join([])` returns an empty string)
```suggestion
    extra_include_paths_str = " ".join(
        [f"-I{include}" for include in extra_include_paths] if extra_include_paths else []
    )
```

File: torch/utils/cpp_extension.py [hunk 3]
```diff
@@ -2344,9 +2347,8 @@ def sanitize_flags(flags):
         # --generate-dependencies-with-compile was added in CUDA 10.2.
         # Compilation will work on earlier CUDA versions but header file
         # dependencies are not correctly computed.
-        required_cuda_version = packaging.version.parse('11.0')
-        has_cuda_version = torch.version.cuda is not None
-        if has_cuda_version and packaging.version.parse(torch.version.cuda) >= required_cuda_version:
+        required_cuda_version = '11.0'
+        if torch.version.cuda is not None and TorchVersion(torch.version.cuda) >= required_cuda_version:
```
Comment (malfet) [inline torch/utils/cpp_extension.py:2351]: TODO (just remove this one, perhaps in followup PR)

**Review by malfet — APPROVED** [review]

File: torch/utils/cpp_extension.py (hunk 2 above)
Comment (malfet) [reply torch/utils/cpp_extension.py:1455]: Ok, as CI is green, let's land it as is, will submit tiny BE fixes later

**Pull request closed without merging**